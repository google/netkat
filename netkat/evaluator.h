// Copyright 2024 The NetKAT authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// -----------------------------------------------------------------------------
// File: evaluator.h
// -----------------------------------------------------------------------------
//
// Defines a library of functions for evaluating NetKAT predicates and policies
// on concrete packets.
//
// See go/netkat-hld for more details.

#ifndef GOOGLE_NETKAT_NETKAT_EVALUATOR_H_
#define GOOGLE_NETKAT_NETKAT_EVALUATOR_H_

#include <string>

#include "absl/container/flat_hash_map.h"
#include "absl/container/flat_hash_set.h"
#include "netkat/netkat.pb.h"

namespace netkat {

// A NetKAT packet is a map from field names to their values.
//
// NOTE: This is a simplistic, initial definition of `Packet` that we expect to
// replace in the future. In particular, the value type will change to support
// things like 128-bit IPv6 addresses.
//
// Fields that are not present in the map are assumed to carry an implicit "not
// present" value that is distinct from all explicitly-assignable values.
using Packet = absl::flat_hash_map<std::string, int>;

// Returns true if the given `packet` satisfies the given `predicate`, false
// otherwise.
//
// Note: Uninitialized predicates are considered unsatisfiable.
bool Evaluate(const PredicateProto& predicate, const Packet& packet);

// Returns the output packets produced by running the given policy on the given
// input packet. Treats `Record` (aka `dup`) as no-op and does not keep track of
// packet histories.
//
// Note: Uninitialized policies are considered DENY, returning the empty set.
absl::flat_hash_set<Packet> Evaluate(const PolicyProto& policy,
                                     const Packet& packet);

// Lifts policy evaluation to sets of packets.
absl::flat_hash_set<Packet> Evaluate(
    const PolicyProto& policy, const absl::flat_hash_set<Packet>& packets);

}  // namespace netkat

#endif  // GOOGLE_NETKAT_NETKAT_EVALUATOR_H_
