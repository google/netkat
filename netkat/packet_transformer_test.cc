// Copyright 2025 The NetKAT authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "netkat/packet_transformer.h"

#include <cstdint>
#include <ostream>
#include <utility>

#include "absl/base/no_destructor.h"
#include "absl/container/flat_hash_map.h"
#include "absl/container/flat_hash_set.h"
#include "absl/log/log.h"
#include "absl/strings/str_cat.h"
#include "absl/strings/str_split.h"
#include "absl/strings/string_view.h"
#include "fuzztest/fuzztest.h"
#include "gmock/gmock.h"
#include "gtest/gtest.h"
#include "gutil/status_matchers.h"  // IWYU pragma: keep
#include "netkat/evaluator.h"
#include "netkat/netkat.pb.h"
#include "netkat/netkat_proto_constructors.h"
#include "netkat/packet_set.h"
#include "re2/re2.h"

namespace netkat {

// We use a global manager object to exercise statefulness more deeply across
// test cases. This also enables better pretty printing for debugging, see
// `PrintTo`.
PacketTransformerManager& Manager() {
  static absl::NoDestructor<PacketTransformerManager> manager;
  return *manager;
}

// The default `PacketTransformerHandle` pretty printer sucks! It does not
// have access to the graph structure representing the packet, since that is
// stored in the manager object. Thus, it returns opaque strings like
// "PacketTransformerHandle<123>".
//
// We define this much better override, which GoogleTest gives precedence to.
void PrintTo(const PacketTransformerHandle& transformer, std::ostream* os) {
  *os << Manager().ToString(transformer);
}

namespace {

using ::testing::ContainerEq;
using ::testing::IsEmpty;
using ::testing::Pair;
using ::testing::StartsWith;
using ::testing::UnorderedElementsAre;

// After executing all tests, we check once that no invariants are violated, for
// defense in depth. Checking invariants after each test (e.g. using a fixture)
// would likely not scale and seems overkill.
class CheckPacketTransformerManagerInvariantsOnTearDown
    : public testing::Environment {
 public:
  ~CheckPacketTransformerManagerInvariantsOnTearDown() override = default;
  void SetUp() override {}
  void TearDown() override { ASSERT_OK(Manager().CheckInternalInvariants()); }
};
testing::Environment* const foo_env = testing::AddGlobalTestEnvironment(
    new CheckPacketTransformerManagerInvariantsOnTearDown);

/*--- Basic tests ------------------------------------------------------------*/

TEST(PacketTransformerManagerTest, DenyIsDeny) {
  EXPECT_TRUE(Manager().IsDeny(Manager().Deny()));
  EXPECT_FALSE(Manager().IsAccept(Manager().Deny()));
}

TEST(PacketTransformerManagerTest, AcceptIsAccept) {
  EXPECT_TRUE(Manager().IsAccept(Manager().Accept()));
  EXPECT_FALSE(Manager().IsDeny(Manager().Accept()));
}

TEST(PacketTransformerManagerTest, DenyDoesNotEqualAccept) {
  EXPECT_NE(Manager().Deny(), Manager().Accept());
}

TEST(PacketTransformerManagerTest, AbslStringifyWorksForDeny) {
  EXPECT_THAT(absl::StrCat(Manager().Deny()),
              StartsWith("PacketTransformerHandle"));
}

TEST(PacketTransformerManagerTest, AbslStringifyWorksForAccept) {
  EXPECT_THAT(absl::StrCat(Manager().Accept()),
              StartsWith("PacketTransformerHandle"));
}

TEST(PacketTransformerManagerTest, AbslHashValueWorks) {
  absl::flat_hash_set<PacketTransformerHandle> set = {
      Manager().Deny(),
      Manager().Accept(),
  };
  EXPECT_EQ(set.size(), 2);
}

TEST(PacketTransformerManagerTest, EmptyPolicyCompilesToDeny) {
  EXPECT_TRUE(Manager().IsDeny(Manager().Compile(PolicyProto())));
}

TEST(PacketTransformerManagerTest, RecordPolicyCompilesToAccept) {
  EXPECT_TRUE(Manager().IsAccept(Manager().Compile(RecordProto())));
}

// packet transformer compile should give the same result as
// PacketSetHandle -> OfPacketSetHandle, if PolicyProto is only a Filter.
void CompileIsSameAsOfCompiledPacketSetHandle(PredicateProto predicate) {
  PacketSetHandle set_1 = Manager().GetPacketSetManager().Compile(predicate);
  EXPECT_EQ(Manager().Compile(FilterProto(predicate)),
            Manager().FromPacketSetHandle(set_1));

  // Using a newly constructed PacketSetManager.
  PacketSetManager packet_set_manager;
  PacketSetHandle set_2 = packet_set_manager.Compile(predicate);
  PacketTransformerManager manager(std::move(packet_set_manager));
  EXPECT_EQ(manager.Compile(FilterProto(predicate)),
            manager.FromPacketSetHandle(set_2));
}
FUZZ_TEST(PacketTransformerManagerTest,
          CompileIsSameAsOfCompiledPacketSetHandle);

/*--- Basic compilation and method consistency checks ------------------------*/

TEST(PacketTransformerManagerTest, AcceptCompilesToAccept) {
  EXPECT_EQ(Manager().Compile(AcceptProto()), Manager().Accept());
}

TEST(PacketTransformerManagerTest, DenyCompilesToDeny) {
  EXPECT_EQ(Manager().Compile(DenyProto()), Manager().Deny());
}

void FilterCompilesToFilter(PredicateProto predicate) {
  EXPECT_EQ(Manager().Compile(FilterProto(predicate)),
            Manager().Filter(predicate));
}
FUZZ_TEST(PacketTransformerManagerTest, FilterCompilesToFilter);

void ModificationCompilesToModification(std::string field, int value) {
  EXPECT_EQ(Manager().Compile(ModificationProto(field, value)),
            Manager().Modification(field, value));
}
FUZZ_TEST(PacketTransformerManagerTest, ModificationCompilesToModification);

void UnionCompilesToUnion(PolicyProto left, PolicyProto right) {
  EXPECT_EQ(Manager().Compile(UnionProto(left, right)),
            Manager().Union(Manager().Compile(left), Manager().Compile(right)));
}
FUZZ_TEST(PacketTransformerManagerTest, UnionCompilesToUnion);

void SequenceCompilesToSequence(PolicyProto left, PolicyProto right) {
  EXPECT_EQ(
      Manager().Compile(SequenceProto(left, right)),
      Manager().Sequence(Manager().Compile(left), Manager().Compile(right)));
}
FUZZ_TEST(PacketTransformerManagerTest, SequenceCompilesToSequence);

void IterateCompilesToIterate(PolicyProto iterable) {
  EXPECT_EQ(Manager().Compile(IterateProto(iterable)),
            Manager().Iterate(Manager().Compile(iterable)));
}
FUZZ_TEST(PacketTransformerManagerTest, IterateCompilesToIterate);

/*--- Kleene algebra axioms and equivalences ---------------------------------*/

void UnionIsAssociative(PolicyProto a, PolicyProto b, PolicyProto c) {
  EXPECT_EQ(Manager().Compile(UnionProto(a, UnionProto(b, c))),
            Manager().Compile(UnionProto(UnionProto(a, b), c)));
}
FUZZ_TEST(PacketTransformerManagerTest, UnionIsAssociative);

void UnionIsCommutative(PolicyProto a, PolicyProto b) {
  EXPECT_EQ(Manager().Compile(UnionProto(a, b)),
            Manager().Compile(UnionProto(b, a)));
}
FUZZ_TEST(PacketTransformerManagerTest, UnionIsCommutative);

void UnionDenyIsIdentity(PolicyProto policy) {
  EXPECT_EQ(Manager().Compile(UnionProto(policy, DenyProto())),
            Manager().Compile(policy));
}
FUZZ_TEST(PacketTransformerManagerTest, UnionDenyIsIdentity);

void UnionIsIdempotent(PolicyProto policy) {
  EXPECT_EQ(Manager().Compile(UnionProto(policy, policy)),
            Manager().Compile(policy));
}
FUZZ_TEST(PacketTransformerManagerTest, UnionIsIdempotent);

void SequenceIsAssociative(PolicyProto a, PolicyProto b, PolicyProto c) {
  EXPECT_EQ(Manager().Compile(SequenceProto(a, SequenceProto(b, c))),
            Manager().Compile(SequenceProto(SequenceProto(a, b), c)));
}
FUZZ_TEST(PacketTransformerManagerTest, SequenceIsAssociative);

void SequenceAcceptIsIdentity(PolicyProto policy) {
  EXPECT_EQ(Manager().Compile(SequenceProto(policy, AcceptProto())),
            Manager().Compile(policy));
  EXPECT_EQ(Manager().Compile(SequenceProto(AcceptProto(), policy)),
            Manager().Compile(policy));
}
FUZZ_TEST(PacketTransformerManagerTest, SequenceAcceptIsIdentity);

void SequenceDenyIsAlwaysDeny(PolicyProto policy) {
  EXPECT_TRUE(
      Manager().IsDeny(Manager().Compile(SequenceProto(policy, DenyProto()))));
  EXPECT_TRUE(
      Manager().IsDeny(Manager().Compile(SequenceProto(DenyProto(), policy))));
}
FUZZ_TEST(PacketTransformerManagerTest, SequenceDenyIsAlwaysDeny);

void DistributiveLawsHold(PolicyProto a, PolicyProto b, PolicyProto c) {
  // Left distribution.
  EXPECT_EQ(
      Manager().Compile(SequenceProto(a, UnionProto(b, c))),
      Manager().Compile(UnionProto(SequenceProto(a, b), SequenceProto(a, c))));
  // Right distribution.
  EXPECT_EQ(
      Manager().Compile(SequenceProto(UnionProto(a, b), c)),
      Manager().Compile(UnionProto(SequenceProto(a, c), SequenceProto(b, c))));
}
FUZZ_TEST(PacketTransformerManagerTest, DistributiveLawsHold);

void IterateUnrollOnce(PolicyProto policy) {
  // Left unroll.
  EXPECT_EQ(Manager().Compile(UnionProto(
                AcceptProto(), SequenceProto(policy, IterateProto(policy)))),
            Manager().Compile(IterateProto(policy)));
  // Right unroll.
  EXPECT_EQ(Manager().Compile(UnionProto(
                AcceptProto(), SequenceProto(IterateProto(policy), policy))),
            Manager().Compile(IterateProto(policy)));
}
FUZZ_TEST(PacketTransformerManagerTest, IterateUnrollOnce);

// This test checks that iterate is the least-fixed point on the left and right
// side of a sequence. I.e. that if there is a term x such that x;y (or y;x) is
// smaller than y, then x* is the smallest such term.
void IterateIsLeastFixedPoint(PolicyProto p, PolicyProto q, PolicyProto r) {
  // Left.
  PolicyProto lfp_left_antecedent = UnionProto(q, SequenceProto(p, r));
  if (Manager().Compile(UnionProto(lfp_left_antecedent, r)) ==
      Manager().Compile(r)) {
    PolicyProto left_lfp = SequenceProto(IterateProto(p), q);
    EXPECT_EQ(Manager().Compile(UnionProto(left_lfp, r)), Manager().Compile(r));
  }
  // Right.
  PolicyProto lfp_right_antecedent = UnionProto(p, SequenceProto(q, r));
  if (Manager().Compile(UnionProto(lfp_right_antecedent, q)) ==
      Manager().Compile(q)) {
    PolicyProto right_lfp = SequenceProto(p, IterateProto(r));
    EXPECT_EQ(Manager().Compile(UnionProto(right_lfp, q)),
              Manager().Compile(q));
  }
}
FUZZ_TEST(PacketTransformerManagerTest, IterateIsLeastFixedPoint);

/*--- Tests with concrete protos ---------------------------------------------*/

TEST(PacketTransformerManagerTest, KatchPaperFig5) {
  // (a=5 + b=2);(b:=1 + c=5)
  PolicyProto p = SequenceProto(
      UnionProto(FilterProto(MatchProto("a", 5)),
                 FilterProto(MatchProto("b", 2))),
      UnionProto(ModificationProto("b", 1), FilterProto(MatchProto("c", 5))));

  // (b=1 + c:=4 + a:=1;b:=1)
  PolicyProto q = UnionProto(
      FilterProto(MatchProto("b", 1)),
      UnionProto(
          ModificationProto("c", 4),
          SequenceProto(ModificationProto("a", 1), ModificationProto("b", 1))));

  PacketTransformerHandle p_transformer = Manager().Compile(p);
  PacketTransformerHandle q_transformer = Manager().Compile(q);
  PacketTransformerHandle sequence_transformer =
      Manager().Compile(SequenceProto(p, q));

  EXPECT_EQ(Manager().Sequence(p_transformer, q_transformer),
            sequence_transformer);
}

TEST(PacketTransformerManagerTest, SequenceOfNonLoopProducerConvergesToDeny) {
  // a=1 ; b:=1 ; a:=0
  PolicyProto a_to_b = SequenceProto(
      FilterProto(MatchProto("a", 1)),
      SequenceProto(ModificationProto("b", 1), ModificationProto("a", 0)));

  // b=1 ; b:=0 ; a:=1
  PolicyProto b_to_a = SequenceProto(
      FilterProto(MatchProto("b", 1)),
      SequenceProto(ModificationProto("b", 0), ModificationProto("a", 1)));

  // !(once=1) ; a:=1 ; once:=1
  PolicyProto b_to_a_once =
      SequenceProto(FilterProto(NotProto(MatchProto("once", 1))),
                    SequenceProto(b_to_a, ModificationProto("once", 1)));

  PacketTransformerHandle a_to_b_and_b_to_a_once_transformer =
      Manager().Compile(UnionProto(a_to_b, b_to_a_once));

  PacketTransformerHandle sequenced_transformer2 = Manager().Sequence(
      a_to_b_and_b_to_a_once_transformer, a_to_b_and_b_to_a_once_transformer);

  // Should converge to Deny if sequenced 4 times.
  PacketTransformerHandle sequenced_transformer4 =
      Manager().Sequence(sequenced_transformer2, sequenced_transformer2);

  EXPECT_TRUE(Manager().IsDeny(sequenced_transformer4))
      << "a_to_b_and_b_to_a_once_transformer:\n"
      << Manager().ToString(a_to_b_and_b_to_a_once_transformer)
      << "\nsequenced_transformer4:\n"
      << Manager().ToString(sequenced_transformer4);
}

TEST(PacketTransformerManagerTest, SequenceOfLoopProducerConvergesToNonDeny) {
  // a=1 ; b:=1 ; a:=0
  PolicyProto a_to_b = SequenceProto(
      FilterProto(MatchProto("a", 1)),
      SequenceProto(ModificationProto("b", 1), ModificationProto("a", 0)));

  // b=1 ; b:=0 ; a:=1
  PolicyProto b_to_a = SequenceProto(
      FilterProto(MatchProto("b", 1)),
      SequenceProto(ModificationProto("b", 0), ModificationProto("a", 1)));

  PacketTransformerHandle a_to_b_and_b_to_a_transformer =
      Manager().Compile(UnionProto(a_to_b, b_to_a));

  PacketTransformerHandle sequenced_transformer2 = Manager().Sequence(
      a_to_b_and_b_to_a_transformer, a_to_b_and_b_to_a_transformer);
  PacketTransformerHandle sequenced_transformer4 =
      Manager().Sequence(sequenced_transformer2, sequenced_transformer2);

  EXPECT_FALSE(Manager().IsDeny(sequenced_transformer4));
  EXPECT_EQ(sequenced_transformer2, sequenced_transformer4)
      << "nsequenced_transformer2:\n"
      << Manager().ToString(sequenced_transformer2)
      << "\nsequenced_transformer4:\n"
      << Manager().ToString(sequenced_transformer4);
}

// Tests that a simple sequence of modification then filter same field with
// different value is Deny.
TEST(PacketTransformerManagerTest, ModifyThenFilterDifferentValueIsDeny) {
  // a:=0 ; a=1
  PolicyProto make_false_then_test =
      SequenceProto(ModificationProto("a", 0), FilterProto(MatchProto("a", 1)));

  EXPECT_TRUE(Manager().IsDeny(Manager().Compile(make_false_then_test)))
      << Manager().ToString(Manager().Compile(make_false_then_test));
}

// Tests that a simple sequence of modification then filter same field with
// same value is Modify.
TEST(PacketTransformerManagerTest, ModifyThenFilterSameValueIsModify) {
  // a:=1 ; a=1
  PolicyProto make_true = ModificationProto("a", 1);
  PolicyProto make_true_then_test =
      SequenceProto(make_true, FilterProto(MatchProto("a", 1)));

  EXPECT_EQ(Manager().Compile(make_true_then_test),
            Manager().Compile(make_true));
}

/*--- Tests with packets -----------------------------------------------------*/

TEST(PacketTransformerManagerTest, RunDenyAndAccept) {
  Packet packet = {{"field", 1}};
  Packet original_packet = packet;
  EXPECT_THAT(Manager().Run(Manager().Deny(), packet), IsEmpty());
  EXPECT_EQ(packet, original_packet);
  EXPECT_THAT(Manager().Run(Manager().Accept(), packet),
              UnorderedElementsAre(packet));
  EXPECT_EQ(packet, original_packet);
}

// We expect that any concrete packet that is `Run` through a `policy` gives the
// same result as when it is `Evaluate`d on that policy.
void RunIsSameAsEvaluate(PolicyProto policy, Packet packet) {
  Packet original_packet = packet;
  EXPECT_THAT(Manager().Run(Manager().Compile(policy), packet),
              ContainerEq(Evaluate(policy, packet)));
  EXPECT_EQ(packet, original_packet);
}
FUZZ_TEST(PacketTransformerManagerTest, RunIsSameAsEvaluate);

TEST(PacketTransformerManagerTest, SimpleSequenceRunTest1) {
  // !(once=1) ; a:=1 ; once:=1
  PacketTransformerHandle match_then_modify_transformer = Manager().Compile(
      SequenceProto(FilterProto(NotProto(MatchProto("once", 1))),
                    SequenceProto(ModificationProto("a", 1),
                                  ModificationProto("once", 1))));

  Packet packet_without_once;
  Packet packet_without_once_copy = packet_without_once;
  Packet packet_with_once_1 = {{"once", 1}};
  Packet packet_with_once_1_copy = packet_with_once_1;
  Packet packet_with_once_0 = {{"once", 0}};
  Packet packet_with_once_0_copy = packet_with_once_0;
  Packet expected_packet = {{"once", 1}, {"a", 1}};
  EXPECT_THAT(Manager().Run(match_then_modify_transformer, packet_without_once),
              UnorderedElementsAre(expected_packet));
  EXPECT_EQ(packet_without_once, packet_without_once_copy);
  EXPECT_THAT(Manager().Run(match_then_modify_transformer, packet_with_once_1),
              IsEmpty());
  EXPECT_EQ(packet_with_once_1, packet_with_once_1_copy);

  EXPECT_THAT(Manager().Run(match_then_modify_transformer, packet_with_once_0),
              UnorderedElementsAre(expected_packet));
  EXPECT_EQ(packet_with_once_0, packet_with_once_0_copy);
}

TEST(PacketTransformerManagerTest, SimpleSequenceAndUnionRunTest2) {
  // a=1 ; a:=0
  PacketTransformerHandle check_a = Manager().Compile(SequenceProto(
      FilterProto(MatchProto("a", 1)), ModificationProto("a", 0)));

  // Does `a:=1` exactly once.
  // !(once=1) ; a:=1 ; once:=1
  PacketTransformerHandle a_once = Manager().Compile(SequenceProto(
      FilterProto(NotProto(MatchProto("once", 1))),
      SequenceProto(ModificationProto("a", 1), ModificationProto("once", 1))));

  PacketTransformerHandle check_a_and_a_once_transformer =
      Manager().Union(check_a, a_once);

  Packet packet_without_fields;
  Packet packet_with_once_0 = {{"once", 0}};
  Packet packet_with_once_1 = {{"once", 1}};
  Packet packet_with_a_0 = {{"a", 0}};
  Packet packet_with_a_1 = {{"a", 1}};
  Packet packet_with_a_1_copy = packet_with_a_1;
  Packet expected_packet_a0 = {{"once", 1}, {"a", 0}};
  Packet expected_packet_a1 = {{"once", 1}, {"a", 1}};

  // Test `check_a_and_a_once_transformer`.
  EXPECT_THAT(
      Manager().Run(check_a_and_a_once_transformer, packet_without_fields),
      UnorderedElementsAre(expected_packet_a1));

  EXPECT_THAT(Manager().Run(check_a_and_a_once_transformer, packet_with_once_0),
              UnorderedElementsAre(expected_packet_a1));
  EXPECT_THAT(Manager().Run(check_a_and_a_once_transformer, packet_with_once_1),
              IsEmpty());

  EXPECT_THAT(Manager().Run(check_a_and_a_once_transformer, packet_with_a_0),
              UnorderedElementsAre(expected_packet_a1));
  EXPECT_THAT(Manager().Run(check_a_and_a_once_transformer, packet_with_a_1),
              UnorderedElementsAre(packet_with_a_0, expected_packet_a1));

  // Run the results through again!
  EXPECT_THAT(Manager().Run(check_a_and_a_once_transformer, expected_packet_a1),
              UnorderedElementsAre(expected_packet_a0));
  EXPECT_THAT(Manager().Run(check_a_and_a_once_transformer, packet_with_a_0),
              UnorderedElementsAre(expected_packet_a1));

  // This should be the same as running the original packets through
  // `sequenced_transformer2`.
  PacketTransformerHandle sequenced_transformer2 = Manager().Sequence(
      check_a_and_a_once_transformer, check_a_and_a_once_transformer);
  // Same result as running `expected_packet_a1` again above:
  EXPECT_THAT(Manager().Run(sequenced_transformer2, packet_without_fields),
              UnorderedElementsAre(expected_packet_a0));
  EXPECT_THAT(Manager().Run(sequenced_transformer2, packet_with_once_0),
              UnorderedElementsAre(expected_packet_a0));
  EXPECT_THAT(Manager().Run(sequenced_transformer2, packet_with_a_0),
              UnorderedElementsAre(expected_packet_a0));

  // Union of results from above.
  EXPECT_THAT(Manager().Run(sequenced_transformer2, packet_with_a_1),
              UnorderedElementsAre(expected_packet_a0, expected_packet_a1));
  EXPECT_EQ(packet_with_a_1, packet_with_a_1_copy);

  // Run it again! Note that `expected_packet_a0` would be created on a third
  // run (from `expected_packet_a1`), so we skip that only do the 4th run with
  // `expected_packet_a0`.
  EXPECT_THAT(Manager().Run(check_a_and_a_once_transformer, expected_packet_a0),
              IsEmpty());

  // Should converge to Deny if sequenced 4 times.
  PacketTransformerHandle sequenced_transformer4 =
      Manager().Sequence(sequenced_transformer2, sequenced_transformer2);

  EXPECT_TRUE(Manager().IsDeny(sequenced_transformer4))
      << "sequenced_transformer4:\n"
      << Manager().ToString(sequenced_transformer4);
}

}  // namespace
}  // namespace netkat
