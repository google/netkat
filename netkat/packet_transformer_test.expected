================================================================================
Test case: p := (a=5 + b=2);(b:=1 + c=5). Example from Katch paper Fig 5.
================================================================================
-- STRING ----------------------------------------------------------------------
PacketTransformerHandle<7>:
  PacketFieldHandle<0>:'a' == 5:
    PacketFieldHandle<0>:'a' := 5 -> PacketTransformerHandle<5>
  PacketFieldHandle<0>:'a' == *:
  PacketFieldHandle<0>:'a' == * -> PacketTransformerHandle<6>
PacketTransformerHandle<5>:
  PacketFieldHandle<1>:'b' == *:
    PacketFieldHandle<1>:'b' := 1 -> PacketTransformerHandle<accept>
  PacketFieldHandle<1>:'b' == * -> PacketTransformerHandle<4>
PacketTransformerHandle<6>:
  PacketFieldHandle<1>:'b' == 2:
    PacketFieldHandle<1>:'b' := 1 -> PacketTransformerHandle<accept>
    PacketFieldHandle<1>:'b' := 2 -> PacketTransformerHandle<4>
  PacketFieldHandle<1>:'b' == *:
  PacketFieldHandle<1>:'b' == * -> PacketTransformerHandle<deny>
PacketTransformerHandle<4>:
  PacketFieldHandle<2>:'c' == 5:
    PacketFieldHandle<2>:'c' := 5 -> PacketTransformerHandle<accept>
  PacketFieldHandle<2>:'c' == *:
  PacketFieldHandle<2>:'c' == * -> PacketTransformerHandle<deny>
-- DOT -------------------------------------------------------------------------
digraph {
  4294967294 [label="T" shape=box]
  4294967295 [label="F" shape=box]
  7 [label="a"]
  7 -> 5 [label="a==5; a:=5"]
  7 -> 6 [style=dashed]
  5 [label="b"]
  5 -> 4294967294 [label="b:=1"]
  5 -> 4 [style=dashed]
  6 [label="b"]
  6 -> 4294967294 [label="b==2; b:=1"]
  6 -> 4 [label="b==2; b:=2"]
  6 -> 4294967295 [style=dashed]
  4 [label="c"]
  4 -> 4294967294 [label="c==5; c:=5"]
  4 -> 4294967295 [style=dashed]
}
================================================================================
Test case: q := (b=1 + c:=4 + a:=5;b:=1). Example from Katch paper Fig 5.
================================================================================
-- STRING ----------------------------------------------------------------------
PacketTransformerHandle<16>:
  PacketFieldHandle<0>:'a' == 1:
    PacketFieldHandle<0>:'a' := 1 -> PacketTransformerHandle<13>
  PacketFieldHandle<0>:'a' == *:
    PacketFieldHandle<0>:'a' := 1 -> PacketTransformerHandle<3>
  PacketFieldHandle<0>:'a' == * -> PacketTransformerHandle<15>
PacketTransformerHandle<13>:
  PacketFieldHandle<1>:'b' == 1:
    PacketFieldHandle<1>:'b' := 1 -> PacketTransformerHandle<12>
  PacketFieldHandle<1>:'b' == *:
    PacketFieldHandle<1>:'b' := 1 -> PacketTransformerHandle<accept>
  PacketFieldHandle<1>:'b' == * -> PacketTransformerHandle<9>
PacketTransformerHandle<3>:
  PacketFieldHandle<1>:'b' == *:
    PacketFieldHandle<1>:'b' := 1 -> PacketTransformerHandle<accept>
  PacketFieldHandle<1>:'b' == * -> PacketTransformerHandle<deny>
PacketTransformerHandle<15>:
  PacketFieldHandle<1>:'b' == 1:
    PacketFieldHandle<1>:'b' := 1 -> PacketTransformerHandle<12>
  PacketFieldHandle<1>:'b' == *:
  PacketFieldHandle<1>:'b' == * -> PacketTransformerHandle<9>
PacketTransformerHandle<12>:
  PacketFieldHandle<2>:'c' == *:
    PacketFieldHandle<2>:'c' := 4 -> PacketTransformerHandle<accept>
  PacketFieldHandle<2>:'c' == * -> PacketTransformerHandle<accept>
PacketTransformerHandle<9>:
  PacketFieldHandle<2>:'c' == *:
    PacketFieldHandle<2>:'c' := 4 -> PacketTransformerHandle<accept>
  PacketFieldHandle<2>:'c' == * -> PacketTransformerHandle<deny>
-- DOT -------------------------------------------------------------------------
digraph {
  4294967294 [label="T" shape=box]
  4294967295 [label="F" shape=box]
  16 [label="a"]
  16 -> 13 [label="a==1; a:=1"]
  16 -> 3 [label="a:=1"]
  16 -> 15 [style=dashed]
  13 [label="b"]
  13 -> 12 [label="b==1; b:=1"]
  13 -> 4294967294 [label="b:=1"]
  13 -> 9 [style=dashed]
  3 [label="b"]
  3 -> 4294967294 [label="b:=1"]
  3 -> 4294967295 [style=dashed]
  15 [label="b"]
  15 -> 12 [label="b==1; b:=1"]
  15 -> 9 [style=dashed]
  12 [label="c"]
  12 -> 4294967294 [label="c:=4"]
  12 -> 4294967294 [style=dashed]
  9 [label="c"]
  9 -> 4294967294 [label="c:=4"]
  9 -> 4294967295 [style=dashed]
}
================================================================================
Test case: p;q. Example from Katch paper Fig 5.
================================================================================
-- STRING ----------------------------------------------------------------------
PacketTransformerHandle<21>:
  PacketFieldHandle<0>:'a' == 1:
    PacketFieldHandle<0>:'a' := 1 -> PacketTransformerHandle<18>
  PacketFieldHandle<0>:'a' == 5:
    PacketFieldHandle<0>:'a' := 1 -> PacketTransformerHandle<3>
    PacketFieldHandle<0>:'a' := 5 -> PacketTransformerHandle<20>
  PacketFieldHandle<0>:'a' == *:
    PacketFieldHandle<0>:'a' := 1 -> PacketTransformerHandle<19>
  PacketFieldHandle<0>:'a' == * -> PacketTransformerHandle<18>
PacketTransformerHandle<18>:
  PacketFieldHandle<1>:'b' == 2:
    PacketFieldHandle<1>:'b' := 1 -> PacketTransformerHandle<12>
    PacketFieldHandle<1>:'b' := 2 -> PacketTransformerHandle<17>
  PacketFieldHandle<1>:'b' == *:
  PacketFieldHandle<1>:'b' == * -> PacketTransformerHandle<deny>
PacketTransformerHandle<3>:
  PacketFieldHandle<1>:'b' == *:
    PacketFieldHandle<1>:'b' := 1 -> PacketTransformerHandle<accept>
  PacketFieldHandle<1>:'b' == * -> PacketTransformerHandle<deny>
PacketTransformerHandle<20>:
  PacketFieldHandle<1>:'b' == *:
    PacketFieldHandle<1>:'b' := 1 -> PacketTransformerHandle<12>
  PacketFieldHandle<1>:'b' == * -> PacketTransformerHandle<17>
PacketTransformerHandle<19>:
  PacketFieldHandle<1>:'b' == 2:
    PacketFieldHandle<1>:'b' := 1 -> PacketTransformerHandle<accept>
  PacketFieldHandle<1>:'b' == *:
  PacketFieldHandle<1>:'b' == * -> PacketTransformerHandle<deny>
PacketTransformerHandle<12>:
  PacketFieldHandle<2>:'c' == *:
    PacketFieldHandle<2>:'c' := 4 -> PacketTransformerHandle<accept>
  PacketFieldHandle<2>:'c' == * -> PacketTransformerHandle<accept>
PacketTransformerHandle<17>:
  PacketFieldHandle<2>:'c' == 5:
    PacketFieldHandle<2>:'c' := 4 -> PacketTransformerHandle<accept>
  PacketFieldHandle<2>:'c' == *:
  PacketFieldHandle<2>:'c' == * -> PacketTransformerHandle<deny>
-- DOT -------------------------------------------------------------------------
digraph {
  4294967294 [label="T" shape=box]
  4294967295 [label="F" shape=box]
  21 [label="a"]
  21 -> 18 [label="a==1; a:=1"]
  21 -> 3 [label="a==5; a:=1"]
  21 -> 20 [label="a==5; a:=5"]
  21 -> 19 [label="a:=1"]
  21 -> 18 [style=dashed]
  18 [label="b"]
  18 -> 12 [label="b==2; b:=1"]
  18 -> 17 [label="b==2; b:=2"]
  18 -> 4294967295 [style=dashed]
  3 [label="b"]
  3 -> 4294967294 [label="b:=1"]
  3 -> 4294967295 [style=dashed]
  20 [label="b"]
  20 -> 12 [label="b:=1"]
  20 -> 17 [style=dashed]
  19 [label="b"]
  19 -> 4294967294 [label="b==2; b:=1"]
  19 -> 4294967295 [style=dashed]
  12 [label="c"]
  12 -> 4294967294 [label="c:=4"]
  12 -> 4294967294 [style=dashed]
  17 [label="c"]
  17 -> 4294967294 [label="c==5; c:=4"]
  17 -> 4294967295 [style=dashed]
}
